plugins/outputs/whatap/whatap.go

```diff
@@ -0,0 +1,176 @@
+)
+
+type Whatap struct {
+	License string          `toml:"license"`
+	Servers []string        `toml:"servers"`
+	Pcode   int64           `toml:"project_code"`
+	Timeout config.Duration `toml:"timeout"`
+	Log     telegraf.Logger `toml:"-"`
+	oname   string
+	oid     int32
+	conn    net.Conn
+	hosts   []string
+}
+
+const sampleConfig = `
+  ## You can create a project on the WhaTap site(https://www.whatap.io) 
+  ## to get license, project code and server IP information.
+
+  ## WhaTap license. Required
+  license = "xxxx-xxxx-xxxx"
+
+  ## WhaTap project code. Required
+  project_code = 1111
+
+  ## WhaTap server IP. Required
+  ## Put multiple IPs. ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+  servers = ["tcp://1.1.1.1:6600","tcp://2.2.2.2:6600"]
+
+  ## Connection timeout.
+  # timeout = "60s"
+`
+
+func (w *Whatap) Connect() error {
+	// Change and connect multiple servers sequentially.
+	for _, host := range w.hosts {
+		client, err := net.DialTimeout(netScheme, host, time.Duration(w.Timeout))
+		if err != nil {
+			w.Log.Errorf("connecting to %q failed: %v", host, err)
+			continue
+		}
+		w.conn = client.(*net.TCPConn)
+		w.Log.Info("Connected ", host)
+		return nil
+	}
+	return fmt.Errorf("could not connect to any server")
+}
+
+func (w *Whatap) Close() error {
+	if w.conn == nil {
+		return nil
+	}
+	err := w.conn.Close()
+	w.conn = nil
+	return err
+}
+
+func (w *Whatap) Description() string {
+	return "Plugin to send metrics to a WhaTap server"
+}
+
+func (w *Whatap) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whatap) Write(metrics []telegraf.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+	if w.conn == nil {
+		if err := w.Connect(); err != nil {
+			return err
+		}
+	}
+	// Transform telegraf metrics to whatap data.
+	for _, m := range metrics {
+		p := wpack.NewTagCountPack()
+		p.Pcode = w.Pcode
+		p.Oid = w.oid
+		p.Category = "telegraf_" + m.Name()
+		for k, v := range m.Fields() {
+			p.Put(k, v)
+		}
+		for k, v := range m.Tags() {
+			p.PutTag(k, v)
+		}
+		p.PutTag("oname", w.oname)
+
+		// Convert time to microseconds.
+		p.Time = m.Time().UnixNano() / int64(time.Millisecond)
+
+		dout := wio.NewDataOutputX()
+		dout.WriteShort(p.GetPackType())
+		p.Write(dout)
+		dout.WriteHeader(netSrcAgentOneway, netSrcAgentVersion, w.Pcode,
+			whash.Hash64Str(w.License))
+
+		if err := w.send(dout.ToByteArray()); err != nil {
+			w.Log.Warnf("cannot send data: %v", err)
+			_ = w.Close()
+		}
+	}
+	return nil
+}
+func (w *Whatap) send(sendbuf []byte) (err error) {
+	nbyteleft := len(sendbuf)
+	var pos int
+	var nbytethistime int
+	for 0 < nbyteleft {
+		deadline := time.Now().Add(time.Duration(w.Timeout))
+		if err := w.conn.SetWriteDeadline(deadline); err != nil {
+			return fmt.Errorf("cannot set write deadline: %v", err)
+		}
+
+		nbytethistime, err = w.conn.Write(sendbuf[pos : pos+nbyteleft])
+		if err != nil {
+			return err
+		}
+		nbyteleft -= nbytethistime
+		pos += nbytethistime
+	}
```

Comment:
Wouldn't this be clearer
```suggestion
	for pos := 0; pos < len(sendbuf); {
		deadline := time.Now().Add(time.Duration(w.Timeout))
		if err := w.conn.SetWriteDeadline(deadline); err != nil {
			return fmt.Errorf("cannot set write deadline: %v", err)
		}

		nbytethistime, err := w.conn.Write(sendbuf[pos : len(sendbuf)])
		if err != nil {
			return err
		}
		pos += nbytethistime
	}
```

Reply:
I have modified it according to your suggestion.